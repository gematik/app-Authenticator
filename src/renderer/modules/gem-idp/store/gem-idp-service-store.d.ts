import { OAUTH2_ERROR_TYPE, TCallback } from '@/renderer/modules/gem-idp/type-definitions';

export type IOpenIdConfiguration = {
  authorization_endpoint: string;
  auth_pair_endpoint: string;
  sso_endpoint: string;
  uri_pair: string;
  token_endpoint: string;
  third_party_authorization_endpoint: string;
  uri_disc: string;
  issuer: string;
  jwks_uri: string;
  exp: number;
  iat: number;
  uri_puk_idp_enc: string;
  uri_puk_idp_sig: string;
  subject_types_supported: ['pairwise'];
  id_token_signing_alg_values_supported: ['BP256R1'];
  response_types_supported: ['code'];
  scopes_supported: ['openid', 'e-rezept', 'pairing'];
  response_modes_supported: ['query'];
  grant_types_supported: ['authorization_code'];
  acr_values_supported: ['gematik-ehealth-loa-high'];
  token_endpoint_auth_methods_supported: ['none'];
  code_challenge_methods_supported: ['S256'];
  kk_app_list_uri: string;
};

export type IIdpEncJwk = {
  use: string;
  kid: string;
  kty: string;
  crv: string;
  x: string;
  y: string;
};

export type IJweChallenge = {
  protected: string;
  iv: string;
  ciphertext: string;
  tag: string;
};

export type TGemIdpServiceStore = {
  sid?: string;
  jwsHbaSignature?: string;
  jwsSmcbSignature?: string;
  challengePath: string;
  idpHost: string;
  challenge: string;
  userConsent?: {
    requested_scopes: Record<string, string>;
    requested_claims: Record<string, string>;
  };
  jweChallenge: null | IJweChallenge;
  openIdConfiguration?: IOpenIdConfiguration;
  idpEncJwk?: IIdpEncJwk;
  errorShown: boolean;
  caughtAuthFlowError?: {
    errorCode: string;
    errorDescription?: string;
    oauthErrorType?: OAUTH2_ERROR_TYPE;
  };
  clientId: string;

  callback: TCallback | undefined;
  deeplink: string;
};
