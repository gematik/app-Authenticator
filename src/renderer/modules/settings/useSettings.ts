/*
 * Copyright 2024 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import { inject, InjectionKey } from 'vue';
import { ISettingsRepository } from '@/renderer/modules/settings/repository';

export type ConfigStore = ISettingsRepository;

export const SETTINGS_STORE_KEY: InjectionKey<ConfigStore> = Symbol('SETTINGS_STORE');

interface ConfigStoreProps {
  repository: ISettingsRepository;
}

export const useSettings = (): ConfigStore => {
  const providedSettingsStore = inject(SETTINGS_STORE_KEY, undefined);

  if (providedSettingsStore) {
    return providedSettingsStore;
  }

  throw new Error("Settings store not initialized. Please use 'SettingsProvider' component to initialize");
};

export const createSettingsStore = ({ repository }: ConfigStoreProps): ConfigStore => {
  const save = repository.save;
  const clear = repository.clear;
  const load = repository.load;
  const exist = repository.exist;
  const setWithoutSave = repository.setWithoutSave;
  return { save, clear, load, exist, setWithoutSave };
};
