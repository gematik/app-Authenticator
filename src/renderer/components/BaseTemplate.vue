<!--
  - Copyright 2025, gematik GmbH
  -
  - Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the
  - European Commission â€“ subsequent versions of the EUPL (the "Licence").
  - You may not use this work except in compliance with the Licence.
  -
  - You find a copy of the Licence in the "Licence" file or at
  - https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
  -
  - Unless required by applicable law or agreed to in writing,
  - software distributed under the Licence is distributed on an "AS IS" basis,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
  - In case of changes by gematik find details in the "Readme" file.
  -
  - See the Licence for the specific language governing permissions and limitations under the Licence.
  -
  - *******
  -
  - For additional notes and disclaimer from gematik and in case of changes by gematik find details in the "Readme" file.
  -->

<script lang="ts" setup>
import AuthFlow from '@/renderer/modules/gem-idp/event-listeners/AuthFlow.vue';
import { computed } from 'vue';
import { IS_DEV, PROCESS_ENVS } from '@/constants';
import { useRoute } from 'vue-router';
import { useSettings } from '@/renderer/modules/settings/useSettings';
import packageJson from '../../../package.json';

let version = packageJson.version;

const route = useRoute();

// #!if MOCK_MODE === 'ENABLED'
if (IS_DEV) {
  version = 'dev ';
}
// #!endif

const isActiveAssistant = computed(() => {
  return route.path.startsWith('/config-assistant');
});

const buildVersion = computed(() => {
  // #!if MOCK_MODE === 'ENABLED'
  if (IS_DEV) {
    return packageJson.version;
  }
  // #!endif

  if (PROCESS_ENVS.BRANCH_NAME && PROCESS_ENVS.BRANCH_NAME !== PROCESS_ENVS.TAG_NAME && PROCESS_ENVS.BUILD_NUMBER) {
    return `${PROCESS_ENVS.BRANCH_NAME}-${PROCESS_ENVS.BUILD_NUMBER}`;
  }
  return PROCESS_ENVS.BUILD_NUMBER === true ? ` ${PROCESS_ENVS.BUILD_NUMBER}` : ' snapshot';
});

const { exist } = useSettings();

const settingsSet = computed(() => exist());
</script>

<template>
  <div class="h-full flex flex-col bg-primary text-defaultext">
    <div class="flex flex-row justify-between px-[32px] py-[32px]">
      <div>
        <router-link id="navLogin" to="/" class="p-1.5 mr-[16px] operator:hover bg-primary">
          {{ $t('registration') }}
        </router-link>
        <router-link id="navSettings" to="/settings" class="p-1.5 mr-[16px] operator:hover bg-primary">
          {{ $t('settings') }}
        </router-link>
        <router-link id="navHelp" to="/help" class="p-1.5 mr-[16px] operator:hover bg-primary">
          {{ $t('help') }}
        </router-link>
        <router-link
          v-if="!settingsSet"
          id="navConfigAssistant"
          to="/config-assistant"
          :class="['p-1.5', 'mr-[16px]', 'operator:hover', 'bg-primary', { 'router-link-active': isActiveAssistant }]"
        >
          {{ $t('config_assistant.menu-title') }}
        </router-link>
      </div>
      <img src="@/assets/logo_gematik.svg" class="object-contain" alt="logo" />
    </div>

    <div id="app-content" class="content flex-1 px-[32px] overflow-auto">
      <slot />
    </div>

    <div class="flex flex-row justify-between relative px-[32px] py-[32px]">
      <router-link to="/imprint" class="inline-flex">
        <img src="@/assets/info.svg" class="object-contain mr-[8px]" alt="info icon" />
        <span id="lblImprint" class="text-sm hover:underline">{{ $t('imprint') }}</span>
      </router-link>
      <div id="lblVersion" class="text-sm cursor-default" :title="buildVersion">{{ $t('version', { version }) }}</div>
    </div>

    <AuthFlow />
  </div>
</template>

<style scoped>
@import '../global.css';

#navLogin:hover,
#navSettings:hover,
#navHelp:hover,
#navConfigAssistant:hover {
  background-color: rgb(240, 240, 240);
  border-bottom: rgba(78, 91, 166, 0.5) 2px solid;
}
</style>
