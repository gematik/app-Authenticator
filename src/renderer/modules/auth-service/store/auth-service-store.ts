/*
 * Copyright 2023 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import { ActionContext, Module } from 'vuex';

import { TRootStore } from '@/renderer/store';
import { IAuthServiceStore } from './auth-service-store.d';
import { logger } from '@/renderer/service/logger';
import { KeycloakPluginError, UserfacingError } from '@/renderer/errors/errors';
import getIdpTlsCertificates from '@/renderer/utils/get-idp-tls-certificates';
import { TOidcProtocol2UrlSpec } from '@/@types/common-types';
import { validateAuthResponseData } from '@/renderer/service/idp-response-validators';
import { userAgent } from '@/renderer/utils/utils';

export const INITIAL_AUTH_STORE_STATE = {
  challengePath: '',
  challenge: '',
  authRequestPath: '',
};

const httpsReqConfig = () => ({
  https: {
    certificateAuthority: getIdpTlsCertificates(),
    rejectUnauthorized: true,
  },
  headers: {
    'Content-Type': 'application/json; charset=UTF-8',
    'User-Agent': userAgent,
  },
});

export const authServiceStore: Module<IAuthServiceStore, TRootStore> = {
  namespaced: true,
  state: (): IAuthServiceStore => INITIAL_AUTH_STORE_STATE,
  mutations: {
    setHbaJwsSignature(state: IAuthServiceStore, jwsSignature: string): void {
      state.jwsHbaSignature = jwsSignature;
    },
    setSmcbJwsSignature(state: IAuthServiceStore, jwsSignature: string): void {
      state.jwsSmcbSignature = jwsSignature;
    },
    setSid(state: IAuthServiceStore, sid: string): void {
      state.sid = sid;
    },
    setChallengePath(state: IAuthServiceStore, challengePath: string): void {
      state.challengePath = challengePath;
    },
    setChallenge(state: IAuthServiceStore, challenge: string): void {
      state.challenge = challenge;
    },
    resetStore(state: IAuthServiceStore): void {
      state.authRequestPath = undefined;
      state.jwsHbaSignature = undefined;
      state.jwsSmcbSignature = undefined;
      state.sid = undefined;
      state.challengePath = '';
      state.challenge = '';
    },
    setAuthRequestPath(state: IAuthServiceStore, authReqParameters: TOidcProtocol2UrlSpec): void {
      state.authRequestPath = authReqParameters.authz_path;
    },
  },
  actions: {
    async getChallengeData(context: ActionContext<IAuthServiceStore, TRootStore>): Promise<boolean> {
      const { state, commit } = context;
      logger.debug('state.authRequestPath ' + JSON.stringify(context));
      try {
        //  IDP host and Authorization Path
        const url = state.authRequestPath;
        const idpHost = new URL(<string>url);
        const { data } = await window.api.httpGet(<string>url, { ...httpsReqConfig(), followRedirect: true });

        logger.info('IdP: Challenge data received successfully.');

        // extract IDP Host without Authorization Path
        // Challenge Path:challenge_endpoint received from IDP Keycloak
        const challengePath: string = idpHost.protocol + '//' + idpHost.host + data.challenge_endpoint;

        /**
         * Validates Auth Response and throws error on invalid case
         */
        validateAuthResponseData(data);

        logger.info('IdP: Challenge data validation completed.');

        if (data.challenge_endpoint) {
          commit('setChallengePath', challengePath);
        }
        commit('setSid', data.sid);
        commit('setChallenge', data.challenge);

        return true;
      } catch (e: any) {
        // error comes from validateAuthResponseData, just forward it
        if (e instanceof UserfacingError) {
          throw e;
        }

        const response = e?.response;
        const errorMessage = response?.body?.error || e.message;
        logger.error(`Cannot get challenge from IdP. IdP responds with error: ${errorMessage}`, {
          error: e,
          status: response?.statusCode,
          data: response?.body,
          error_uri: response?.headers?.error_uri,
        });

        throw new KeycloakPluginError('Could not get challenge data for authentication', {
          error: response?.body?.error,
          url: response?.headers['error_uri'] || response?.request?.responseURL,
        });
      }
    },
    async getRedirectUriWithToken(
      context: ActionContext<IAuthServiceStore, TRootStore>,
    ): Promise<{ redirectUri: string; statusCode: number; error_uri?: string }> {
      const { state } = context;
      try {
        const payload = {
          hba: {
            signed_challenge: state.jwsHbaSignature,
          },
          smcb: {
            signed_challenge: state.jwsSmcbSignature,
          },
        };

        const response = await window.api.httpPost(state.challengePath, payload, httpsReqConfig());
        logger.debug(`IdP: Token received successfully. Http Status: ${response.status}`);

        return {
          redirectUri: <string>response?.headers?.['x-callback-location'],
          statusCode: response.status,
          error_uri: <string>response?.headers?.error_uri,
        };
      } catch (e: any) {
        const response = e?.response;
        logger.error('Could not get authorization code from IdP. Details: ', {
          error: e,
          status: response?.statusCode,
          data: response?.body,
          error_uri: response?.headers?.error_uri,
        });
        throw new KeycloakPluginError('Could not get authorization code from IdP. Details: ', {
          statusCode: response?.statusCode,
          error: response?.body?.error,
          url: response?.headers?.error_uri,
        });
      }
    },
  },
};
