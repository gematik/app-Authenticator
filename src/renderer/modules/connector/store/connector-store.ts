/*
 * Copyright 2023 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import { ActionContext, Module } from 'vuex';
import { TRootStore } from '@/renderer/store';
import ConnectorConfig from '@/renderer/modules/connector/connector_impl/connector-config';
import { launch as authSignLauncher } from '@/renderer/modules/connector/connector_impl/auth-sign-launcher';
import { launch as certificateReaderLauncher } from '@/renderer/modules/connector/connector_impl/certificate-reader-launcher';
import { getConfig } from '@/renderer/utils/get-configs';
import { launch as getCardTerminalsLauncher } from '@/renderer/modules/connector/connector_impl/get-card-terminals-launcher';
import { launch as getCardLauncher } from '@/renderer/modules/connector/connector_impl/get-cards-launcher';

import { logger } from '@/renderer/service/logger';
import { validateCardCertificate, validateSignedChallenge } from '@/renderer/modules/connector/services';
import { ConnectorError, UserfacingError } from '@/renderer/errors/errors';

import * as pinVerifier from '@/renderer/modules/connector/connector_impl/verify-pin-launcher';
import * as pinChecker from '@/renderer/modules/connector/connector_impl/check-pin-status';
import { TCardData, TConnectorStore } from '@/renderer/modules/connector/type-definitions';
import { ECardTypes } from '@/renderer/modules/connector/ECardTypes';
import { ERROR_CODES } from '@/error-codes';
import { IPC_OGR_IDP_START_EVENT } from '@/constants';
import { OgrJwsOptions } from '@/renderer/modules/auth-service/sign-feature';
import { GemIdpJwsOptions } from '@/renderer/modules/gem-idp/sign-feature/cidp-sign-options';

/* @if MOCK_MODE == 'ENABLED' */
import {
  MOCK_CONNECTOR_CERTS_CONFIG,
  MOCK_CONNECTOR_CONFIG,
} from '@/renderer/modules/connector/connector-mock/mock-config';
import { MOCK_CARD_PIN_STATUS, MOCK_CARD_TERMINALS } from '@/renderer/modules/connector/connector-mock/mock-constants';
import {
  MockCIdpJWSOptions,
  MockOgrJWSOptions,
} from '@/renderer/modules/connector/connector-mock/jws-jose-tools/jws-tool-helper';
import { MockJwsSignature } from '@/renderer/modules/connector/connector-mock/jws-jose-tools/mock-jws-signature';
import { readMockCertificate } from '@/renderer/modules/connector/connector-mock/mock-utils';
/* @endif */

const base64url = require('base64url');
const PIN_STATUS_VERIFIED = 'VERIFIED';
const PIN_STATUS_VERIFIABLE = 'VERIFIABLE';

const INITIAL_CARDS_STATE = {
  [ECardTypes.HBA]: undefined,
  [ECardTypes.SMCB]: undefined,
};

async function paramsToSignChallenge(
  cardType: ECardTypes,
  isDefaultFlow: boolean,
  context: ActionContext<TConnectorStore, TRootStore>,
  challenge: string,
  sid: string,
) {
  const certificate = <string>context.state.cards[cardType]?.certificate;

  const jwsUtil = isDefaultFlow
    ? new OgrJwsOptions(challenge, certificate, cardType).initializeJwsOptions(sid)
    : new GemIdpJwsOptions(challenge, certificate, cardType).initializeJwsOptions();

  let hashedToSignInputString;
  try {
    hashedToSignInputString = await jwsUtil.hashedChallenge;
    ConnectorConfig.setAuthSignParameter({
      ...ConnectorConfig.authSignParameter,
      base64data: <string>hashedToSignInputString,
    });
  } catch (err) {
    logger.error(`Could not create signed input. Error: ${err.message}`);
    throw new UserfacingError('JWS hashing failed', err.message, ERROR_CODES.AUTHCL_0003);
  }
  return { jwsUtil };
}

export const connectorStore: Module<TConnectorStore, TRootStore> = {
  namespaced: true,
  state: {
    cards: { ...INITIAL_CARDS_STATE },
    flowType: undefined,
  },
  mutations: {
    setHbaCardData(state: TConnectorStore, cardData: TCardData): void {
      const hbaCardData = { ...state.cards.HBA, ...cardData };
      state.cards = { ...state.cards, [ECardTypes.HBA]: hbaCardData };
    },
    setSmcbCardData(state: TConnectorStore, cardData: TCardData): void {
      const smcbCardData = { ...state.cards[ECardTypes.SMCB], ...cardData };
      state.cards = { ...state.cards, [ECardTypes.SMCB]: smcbCardData };
    },
    setTerminals(state: TConnectorStore, terminals: any): void {
      state.terminals = terminals;
    },
    setFlowType(state: TConnectorStore, flowType: string): void {
      state.flowType = flowType;
    },

    resetStore(state: TConnectorStore): void {
      logger.debug('resetStore');
      state.cards = { ...INITIAL_CARDS_STATE };
      state.terminals = undefined;
      state.flowType = undefined;
    },
  },
  actions: {
    async getSignedAuthChallenge(context: ActionContext<TConnectorStore, TRootStore>, data): Promise<string> {
      const { cardType } = data as {
        cardType: ECardTypes;
      };

      const isDefaultFlow: boolean = this.state.connectorStore.flowType == IPC_OGR_IDP_START_EVENT;
      const challenge = isDefaultFlow ? this.state.authServiceStore.challenge : this.state.gemIdpServiceStore.challenge;

      const sid = <string>this.state.authServiceStore.sid;
      const { jwsUtil } = await paramsToSignChallenge(cardType, isDefaultFlow, context, challenge, sid);
      let jwsSignature;
      /* @if MOCK_MODE == 'ENABLED' */
      if (getConfig(MOCK_CONNECTOR_CONFIG).value) {
        const jwsHelper = isDefaultFlow
          ? new MockOgrJWSOptions(cardType, challenge, sid)
          : new MockCIdpJWSOptions(cardType, challenge);
        try {
          jwsSignature = await new MockJwsSignature(cardType).createJws(
            jwsHelper.getPayload(),
            jwsHelper.getProtectedHeader(),
          );
          logger.debug('jwsSignature: ', jwsSignature);
        } catch (err) {
          logger.error('jwsSignature not created: ', err.message);
        }
      } else {
        /* @endif */
        try {
          const cardData = <TCardData>context.state.cards[cardType];
          logger.debug('cardData', JSON.stringify(cardData));
          const signedChallengeFromConnector = await authSignLauncher(
            cardData.cardHandle,
            <string>context.state.flowType,
          );
          const signature = base64url.fromBase64(signedChallengeFromConnector);
          jwsSignature = `${jwsUtil.header}.${jwsUtil.payload}.${signature}`;
          logger.debug('JWS signature created: ', jwsSignature);
        } catch (err) {
          logger.error(`Could not get signed challenge for authentication: ${err.message}`);
          if (err instanceof ConnectorError || err instanceof UserfacingError) {
            throw err;
          }
          throw new UserfacingError("JWS Couldn't be created", err.message, ERROR_CODES.AUTHCL_1117);
        }
        /* @if MOCK_MODE == 'ENABLED' */
      }
      /* @endif */

      if (!jwsSignature || !validateSignedChallenge(jwsSignature)) {
        logger.error('Could not get valid JWS');
        throw new UserfacingError(
          'JWS signature is not valid',
          'JWS signature missing or invalid',
          ERROR_CODES.AUTHCL_0004,
        );
      }

      logger.debug(`A JWS signature fetched for card type ${cardType}.`);

      return jwsSignature;
    },
    getCardCertificate: async function (
      context: ActionContext<TConnectorStore, TRootStore>,
      cardType: ECardTypes,
    ): Promise<void> {
      let certificate = '';
      let certificateChain;
      const cardHandle = context.state.cards[cardType]?.cardHandle;

      try {
        /* @if MOCK_MODE == 'ENABLED' */
        if (getConfig(MOCK_CONNECTOR_CONFIG).value) {
          logger.warn('Mock connector configuration is enabled! Using mocked certificate chain!');
          certificateChain =
            cardType === ECardTypes.HBA
              ? readMockCertificate(MOCK_CONNECTOR_CERTS_CONFIG.HBA_CERT, true)
              : readMockCertificate(MOCK_CONNECTOR_CERTS_CONFIG.SMCB_CERT, true);
        } else {
          /* @endif */
          certificateChain = await certificateReaderLauncher(<string>cardHandle);

          /* @if MOCK_MODE == 'ENABLED' */
        }
        /* @endif */
        certificate = certificateChain;
      } catch (err) {
        logger.error(`Get certificate error: ${err.message}`);
        throw err;
      }

      if (!certificate || !validateCardCertificate(certificate)) {
        throw new UserfacingError(
          'Invalid card certificate',
          `Card certificate for ${cardType} not found or not valid`,
          ERROR_CODES.AUTHCL_1111,
        );
      }

      if (cardType === ECardTypes.HBA) {
        context.commit('setHbaCardData', { certificate: certificate });
      } else {
        context.commit('setSmcbCardData', { certificate: certificate });
      }
    },
    /**
     * Searches for connected card terminals in the network
     * Error handling happens in the GemIdpAuthFlowProcess.vue
     * @param context
     */
    async getCardTerminals(context: ActionContext<TConnectorStore, TRootStore>): Promise<void> {
      let cardTerminals;
      /* @if MOCK_MODE == 'ENABLED' */
      if (getConfig(MOCK_CONNECTOR_CONFIG).value) {
        logger.warn('Mock connector configuration is enabled! Using mocked CardTerminals!');
        cardTerminals = MOCK_CARD_TERMINALS;
      } else {
        /* @endif */
        cardTerminals = await getCardTerminalsLauncher();

        /* @if MOCK_MODE == 'ENABLED' */
      }
      /* @endif */
      context.commit('setTerminals', cardTerminals);
    },
    async getCardHandle(context: ActionContext<TConnectorStore, TRootStore>, cardType: ECardTypes): Promise<void> {
      let cardHandle, response;

      /* @if MOCK_MODE == 'ENABLED' */
      if (!getConfig(MOCK_CONNECTOR_CONFIG).value) {
        /* @endif */
        response = await getCardLauncher(cardType);
        cardHandle = response?.cardHandle;
        /* @if MOCK_MODE == 'ENABLED' */
      }
      /* @endif */

      logger.debug(`cardHandle for ${cardType.toUpperCase()}-card: ${cardHandle}`);
      if (cardType === ECardTypes.HBA) {
        context.commit('setHbaCardData', { ...response, cardHandle });
      } else {
        context.commit('setSmcbCardData', { ...response, cardHandle });
        logger.debug('setSmcbCardData: response:', response, ' cardHandle:', cardHandle);
      }
    },

    async checkPinStatus(context: ActionContext<TConnectorStore, TRootStore>, cardType: ECardTypes): Promise<boolean> {
      /* @if MOCK_MODE == 'ENABLED' */
      if (getConfig(MOCK_CONNECTOR_CONFIG).value) {
        return true;
      }
      /* @endif */
      let pinStatusResult;
      const cardHandle = context.state.cards[cardType]?.cardHandle;

      try {
        /* @if MOCK_MODE == 'ENABLED' */
        if (getConfig(MOCK_CONNECTOR_CONFIG).value) {
          logger.warn('Mock connector configuration is enabled! Using mocked card PIN status!');
          pinStatusResult = MOCK_CARD_PIN_STATUS;
        } else {
          /* @endif */
          pinStatusResult = await pinChecker.getPinStatus(cardType, <string>cardHandle);

          /* @if MOCK_MODE == 'ENABLED' */
        }
        /* @endif */
        logger.debug(
          `PIN status for ${cardType.toUpperCase()}-card with CardHandle:${cardHandle} is: ${
            pinStatusResult.pinStatus
          }`,
        );
        if (cardType === ECardTypes.HBA) {
          context.commit('setHbaCardData', { pinStatus: pinStatusResult.pinStatus, cardType: ECardTypes.HBA });
        } else {
          context.commit('setSmcbCardData', {
            pinStatus: pinStatusResult.pinStatus,
            cardType: ECardTypes.SMCB,
          });
        }

        return pinStatusResult.pinStatus === PIN_STATUS_VERIFIED;
      } catch (err) {
        logger.error(`Cannot check PIN status due to : ${err.message}`);
        throw err;
      }
    },
    verifyPin: async function (
      context: ActionContext<TConnectorStore, TRootStore>,
      cardType: ECardTypes,
    ): Promise<void> {
      /* @if MOCK_MODE == 'ENABLED' */
      if (getConfig(MOCK_CONNECTOR_CONFIG).value) {
        context.commit('setHbaCardData', { pinStatus: 'VERIFIED' });
        context.commit('setSmcbCardData', { pinStatus: 'VERIFIED' });
      } else {
        /* @endif */
        const cardData = context.state.cards[cardType];
        try {
          if (cardData?.pinStatus === PIN_STATUS_VERIFIABLE) {
            await pinVerifier.launch(context.state.terminals, cardData);
          }
          const pinStatusResult = await pinChecker.getPinStatus(cardType, <string>cardData?.cardHandle);

          if (cardType === ECardTypes.HBA) {
            context.commit('setHbaCardData', { pinStatus: pinStatusResult.pinStatus });
          } else {
            context.commit('setSmcbCardData', { pinStatus: pinStatusResult.pinStatus });
          }
        } catch (err) {
          logger.error(`Cannot VerifyPIN because: ${err.error || err.message}`);
          throw err;
        }
        /* @if MOCK_MODE == 'ENABLED' */
      }
      /* @endif */
    },
  },
};
