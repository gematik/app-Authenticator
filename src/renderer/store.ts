/*
 * Copyright 2025, gematik GmbH
 *
 * Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the
 * European Commission â€“ subsequent versions of the EUPL (the "Licence").
 * You may not use this work except in compliance with the Licence.
 *
 * You find a copy of the Licence in the "Licence" file or at
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
 * In case of changes by gematik find details in the "Readme" file.
 *
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 *
 * ******
 *
 * For additional notes and disclaimer from gematik and in case of changes by gematik find details in the "Readme" file.
 */

import { createStore } from 'vuex';
import { connectorStore } from '@/renderer/modules/connector/store/connector-store';
import { TConnectorStore } from '@/renderer/modules/connector/type-definitions';
import { TIdpServiceStore } from '@/renderer/modules/gem-idp/store/idp-service-store.d';
import { idpServiceStore } from '@/renderer/modules/gem-idp/store/idp-service-store';

export type TRootStore = {
  connectorStore: TConnectorStore;
  idpServiceStore: TIdpServiceStore;
  settingsIsSet: boolean;
  showLoadingSpinner: boolean;
};

export default createStore({
  state: (): TRootStore => {
    return {
      showLoadingSpinner: false,
    } as TRootStore;
  },
  mutations: {
    setShowLoadingSpinner(state, newState: boolean) {
      state.showLoadingSpinner = newState;
    },
  },
  actions: {},
  modules: {
    connectorStore,
    idpServiceStore,
  },
});
