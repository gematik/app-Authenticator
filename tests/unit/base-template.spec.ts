/*
 * Copyright 2025, gematik GmbH
 *
 * Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the
 * European Commission â€“ subsequent versions of the EUPL (the "Licence").
 * You may not use this work except in compliance with the Licence.
 *
 * You find a copy of the Licence in the "Licence" file or at
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
 * In case of changes by gematik find details in the "Readme" file.
 *
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

import { mount, RouterLinkStub } from '@vue/test-utils';
import { useRoute } from 'vue-router';
import BaseTemplate from '@/renderer/components/BaseTemplate.vue';
import i18n from '@/renderer/i18n';
import store from '@/renderer/store';
import * as settings from '@/renderer/modules/settings/useSettings';

jest.mock('vue-router', () => ({
  useRoute: jest.fn(),
}));

jest.spyOn(settings, 'useSettings').mockReturnValue({
  save: jest.fn(),
  clear: jest.fn(),
  load: jest.fn(),
  exist: jest.fn(),
  setWithoutSave: jest.fn(),
});

// Type assertion to treat useRoute as a Jest mock function
const mockedUseRoute = useRoute as jest.Mock;

describe('base template screen', () => {
  beforeEach(() => {
    mockedUseRoute.mockReset();
  });

  it('renders with settings button', async () => {
    const mockRoute = {
      path: '/mock-path',
      name: 'mock-route',
      params: {},
      query: {},
    };
    mockedUseRoute.mockReturnValue(mockRoute);

    const wrapper = mount(BaseTemplate, {
      global: {
        plugins: [store, i18n],
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
