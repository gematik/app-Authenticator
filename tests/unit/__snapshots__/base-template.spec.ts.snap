// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base template screen render with settings button 1`] = `
<div
  class="h-full flex flex-col bg-primary text-defaultext"
>
  <div
    class="flex flex-row justify-between px-[32px] py-[32px]"
  >
    <div>
      <router-link
        class="mr-[16px]"
        id="navLogin"
        to="/"
      >
        Anmeldung
      </router-link>
      <router-link
        id="navSettings"
        to="/settings"
      >
        Einstellungen
      </router-link>
    </div>
    <img
      alt="logo"
      class="object-contain"
    />
  </div>
  <div
    class="content flex-1 px-[32px] overflow-auto"
    id="app-content"
  >
    
    
  </div>
  <div
    class="flex flex-row justify-between relative px-[32px] py-[32px]"
  >
    <router-link
      class="inline-flex"
      to="/imprint"
    >
      <img
        alt="info icon"
        class="object-contain mr-[8px]"
      />
      <span
        class="text-sm hover:underline"
        id="lblImprint"
      >
        Impressum
      </span>
    </router-link>
    <div
      class="text-sm cursor-default"
      id="lblVersion"
      title="4.2.1"
    >
      Version dev 
    </div>
  </div>
  
  <div />
  <!--v-if-->
  
</div>
`;
